src/Symfony/Component/Utf8/Bytes.php

```diff
@@ -0,0 +1,526 @@
+            throw new InvalidArgumentException('Passing an empty delimiter is not supported by this method. Use getIterator() method instead.');
+        }
+
+        $chunks = array();
+        foreach (explode($delimiter, $this->string, $limit ?: PHP_INT_MAX) as $i => $string) {
+            $chunks[$i] = $chunk = clone $this;
+            $chunk->string = $string;
+        }
+
+        return $chunks;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function toLowerCase(): self
+    {
+        try {
+            $current = setlocale(LC_CTYPE, '0');
+            setlocale(LC_CTYPE, 'C');
+
+            $result = clone $this;
+            $result->string = strtolower($this->string);
+        } finally {
+            setlocale(LC_CTYPE, $current);
+
+            return $result;
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function toUpperCase(): self
+    {
+        try {
+            $current = setlocale(LC_CTYPE, '0');
+            setlocale(LC_CTYPE, 'C');
+
+            $result = clone $this;
+            $result->string = strtoupper($this->string);
+        } finally {
+            setlocale(LC_CTYPE, $current);
+
+            return $result;
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function toUpperCaseFirst($allWords = false): self
+    {
+        try {
+            $current = setlocale(LC_CTYPE, '0');
+            setlocale(LC_CTYPE, 'C');
+
+            $result = clone $this;
+            $result->string = $allWords ? ucwords($this->string) : ucfirst($this->string);
+        } finally {
+            setlocale(LC_CTYPE, $current);
+
+            return $result;
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function toFoldedCase(bool $full = true): self
+    {
+        return $this->toLowerCase();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function substr(int $start = 0, int $length = null): self
+    {
+        $result = clone $this;
+        $result->string = substr($this->string, $start, $length);
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function trim(string $charsList = null): self
+    {
+        $result = clone $this;
+        $result->string = trim($this->string, $charsList ?: " \t\n\r\0\x0B");
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function trimLeft(string $charsList = null): self
+    {
+        $result = clone $this;
+        $result->string = ltrim($this->string, $charsList ?: " \t\n\r\0\x0B");
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function trimRight(string $charsList = null): self
+    {
+        $result = clone $this;
+        $result->string = rtrim($this->string, $charsList ?: " \t\n\r\0\x0B");
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getIterator(int $maxChunkLength = 1): \Generator
```

Comment:
the actual contract should be `Traversable` or `Iterator` here, not `Generator` IMO. the fact that we use a generator to implement the iterator should be an implementation detail